File: java/src/java/com/google/mr4c/config/algorithm/AlgorithmConfig.java
/**
  * Copyright 2014 Google Inc. All rights reserved.
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

package com.google.mr4c.config.algorithm;

import com.google.mr4c.algorithm.AlgorithmType;

import java.util.ArrayList;
import java.util.List;

public class AlgorithmConfig {

	private String name;
	private String description;
	private AlgorithmType type;
	private String artifact;
	private List<String> inputs = new ArrayList<String>();
	private List<String> outputs = new ArrayList<String>();
	private List<String> optionalInputs = new ArrayList<String>();
	private List<DimensionConfig> dimensions = new ArrayList<DimensionConfig>();
	private List<String> extras = new ArrayList<String>();


	// for gson	
	private AlgorithmConfig() {}

	public AlgorithmConfig(String name, String description, AlgorithmType type, String artifact) {
		this.name = name;
		this.description = description;
		this.type = type;
		this.artifact = artifact;
	}

	public String getName() {
		return this.name;
	}

	public String getDescription() {
		return this.description;
	}

	public AlgorithmType getType() {
		return this.type;
	}

	public String getArtifact() {
		return this.artifact;
	}

	public List<String> getAllInputDatasets() {
		List<String> result = new ArrayList<String>();
		result.addAll(this.inputs);
		result.addAll(this.optionalInputs);
		return result;
	}

	public List<String> getRequiredInputDatasets() {
		return this.inputs;
	}

	public List<String> getOptionalInputDatasets() {
		return this.optionalInputs;
	}

	public void addInputDataset(String input, boolean optional) {
		if ( optional ) {
			this.optionalInputs.add(input);
		} else {
			this.inputs.add(input);
		}
	}

	public List<String> getOutputDatasets() {
		return this.outputs;
	}
		
	public void addOutputDataset(String output) {
		this.outputs.add(output);
	}

	public List<DimensionConfig> getDimensions() {
		return this.dimensions;
	}
		
	public void addDimension(DimensionConfig dim) {
		this.dimensions.add(dim);
	}

	public List<String> getExtras() {
		return this.extras;
	}
		
	public void addExtras(String extra) {
		this.extras.add(extra);
	}

	public boolean equals(Object obj) {
		if ( this==obj ) return true;
		if ( !obj.getClass().equals(this.getClass()) ) return false;
		AlgorithmConfig config = (AlgorithmConfig) obj;
		if ( !name.equals(config.name) ) return false;
		if ( !description.equals(config.description) ) return false;
		if ( !type.equals(config.type) ) return false;
		if ( !artifact.equals(config.artifact) ) return false;
		if ( !inputs.equals(config.inputs) ) return false;
		if ( !outputs.equals(config.outputs) ) return false;
		if ( !optionalInputs.equals(config.optionalInputs) ) return false;
		if ( !dimensions.equals(config.dimensions) ) return false;
		if ( !extras.equals(config.extras) ) return false;
		return true; 
	}

	public int hashCode() {
		return name.hashCode() +
			description.hashCode() +
			type.hashCode() +
			artifact.hashCode() +
			inputs.hashCode() +
			outputs.hashCode();
	}

}

File: native/src/cpp/impl/algorithm/AlgorithmConfig.cpp
/**
  * Copyright 2014 Google Inc. All rights reserved.
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <stdexcept>

#include "algorithm/algorithm_api.h"
#include "dataset/dataset_api.h"
#include "keys/keys_api.h"
#include "metadata/metadata_api.h"
#include "util/util_api.h"

namespace MR4C {

class AlgorithmConfigImpl {

	friend class AlgorithmConfig;

	private:

		std::map<std::string,std::string> m_params;
		std::set<std::string> m_paramNames;

		AlgorithmConfigImpl() {}

		AlgorithmConfigImpl(const std::map<std::string,std::string>& params) {
			initFrom(params);
		}

		AlgorithmConfigImpl(const AlgorithmConfigImpl& config) {
			initFrom(config);
		}

		void initFrom(const AlgorithmConfigImpl& config) {
			initFrom(config.m_params);
		}

		void initFrom(const std::map<std::string,std::string>& params) {
			m_params = params;
			m_paramNames = keySet(params);
		}

	
		std::set<std::string> getAllParamNames() const {
			return m_paramNames;
		}

		bool hasConfigParam(const std::string& name) const {
			return m_params.count(name)!=0;
		}

		std::string getConfigParam(const std::string& name) const {
			if ( !hasConfigParam(name) ) {
				MR4C_THROW( std::invalid_argument, "Parameter [" << name << "] not available in AlgorithmConfig");
			}
			return m_params.find(name)->second;
		}

		bool getConfigParamAsBoolean(const std::string& name) const {
			return Primitive::fromString<bool>(getConfigParam(name));
		}

		int getConfigParamAsInt(const std::string& name) const {
			return Primitive::fromString<int>(getConfigParam(name));
		}

		double getConfigParamAsDouble(const std::string& name) const {
			return Primitive::fromString<double>(getConfigParam(name));
		}

		bool operator==(const AlgorithmConfigImpl& config) const {
			return m_params==config.m_params && m_paramNames==m_paramNames;
		}

		~AlgorithmConfigImpl() {} 

};

AlgorithmConfig::AlgorithmConfig() {
	m_impl = new AlgorithmConfigImpl();
}

AlgorithmConfig::AlgorithmConfig(const std::map<std::string,std::string>& params) {
	m_impl = new AlgorithmConfigImpl(params);
}

AlgorithmConfig::AlgorithmConfig(const AlgorithmConfig& config) {
	m_impl = new AlgorithmConfigImpl(*config.m_impl);
}

std::set<std::string> AlgorithmConfig::getAllParamNames() const {
	return m_impl->getAllParamNames();
}

bool AlgorithmConfig::hasConfigParam(const std::string& name) const {
	return m_impl->hasConfigParam(name);
}

std::string AlgorithmConfig::getConfigParam(const std::string& name) const {
	return m_impl->getConfigParam(name);
}

bool AlgorithmConfig::getConfigParamAsBoolean(const std::string& name) const {
	return m_impl->getConfigParamAsBoolean(name);
}

int AlgorithmConfig::getConfigParamAsInt(const std::string& name) const {
	return m_impl->getConfigParamAsInt(name);
}

double AlgorithmConfig::getConfigParamAsDouble(const std::string& name) const {
	return m_impl->getConfigParamAsDouble(name);
}

bool AlgorithmConfig::operator==(const AlgorithmConfig& config) const {
	return *m_impl==*config.m_impl;
}

bool AlgorithmConfig::operator!=(const AlgorithmConfig& config) const {
	return !operator==(config);
}

AlgorithmConfig& AlgorithmConfig::operator=(const AlgorithmConfig& config) {
	m_impl->initFrom(*config.m_impl);
	return *this;
}

AlgorithmConfig::~AlgorithmConfig() {
	delete m_impl;
}

}
